File: unix_password_cracker.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-

'''
unix_password_cracker.py
    crack unix password using dictonary
'''

import sys
import os
import crypt
import argparse

ls = os.linesep

class Cracker(object):
    def __init__(self, password_file, dictonary_file):
        self.password_file = password_file
        self.dictonary_file = dictonary_file
        
    def crack_password(self, hashed_password):
        '''
        for each hashed password in password.txt, trying password in dictonary
        file until nothing found.
        '''

        try:
            salt = hashed_password[:12]  # first 12-character of hashed_password
            with open(self.dictonary_file, 'r') as f:
                for line in f:
                    password = line.strip(ls)
                    if crypt.crypt(password, salt) == hashed_password:
                        print(' [+] Password FOUND: {0}'.format(password))
                        return
                print(' [-] Password NOT found.')
        except Exception as e:
            print('ERROR: {0}'.format(e))
            sys.exit(1)
    
    def try_password(self):
        '''
        crack hashed password for each user in password
        '''

        try:
            with open(self.password_file, 'r') as f:
                for line in f:
                    username = line.strip(ls).split(':')[0]
                    hashed_password = line.strip(ls).split(':')[1]
                    if hashed_password != '*' and hashed_password != '!':
                        print('[*] Cracking password for {0}...'.format(username))
                        self.crack_password(hashed_password)
        except Exception as e:
            print('ERROR: {0}.'.format(e))
            sys.exit(1)

    def run(self):
        '''
        start cracking program
        '''

        self.try_password()

def main():
    parser = argparse.ArgumentParser(prog = sys.argv[0], description='Unix password cracker')
    parser.add_argument('-f', dest = 'password', help = 'specify password file')
    parser.add_argument('-d', dest = 'dictonary', help = 'dictonary used to crack password file')
    args = parser.parse_args()
    if args.password == None or args.dictonary == None:
        parser.print_usage()
        sys.exit(0)
    cracker = Cracker(args.password, args.dictonary)
    cracker.run()

if __name__ == '__main__':
    main()

File: unix_password_cracker.c
#ifndef _XOPEN_SOURCE
#define _XOPEN_SOURCE
#endif
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <unistd.h>
#include <ctype.h>

char *crack(char *password, char *dictonary);
void get_user_name_and_hashed_password(char *s, char *user, char *hashed_password);
static inline void usage(char *prog);

int main(int argc, char **argv)
{
    char *password_file = NULL;
    char *dictonary_file = NULL;
    char *optstr = "f:d:h";
    int c;

    while ((c = getopt(argc, argv, optstr)) != -1)
    {
        switch (c)
        {
        case 'f':
            password_file = optarg;
            break;
        case 'd':
            dictonary_file = optarg;
            break;
        case 'h':
            usage(*argv);
            break;
        default:
            abort();
        }
    }

    if (argc < 5)
    {
        usage(*argv);
        exit(-1);
    }

    crack(password_file, dictonary_file);

    return 0;
}

char *crack(char *password_file, char *dictonary_file)
{
    FILE *password_fp = fopen(password_file, "r");
    FILE *dictonary_fp = fopen(dictonary_file, "r");
    char user[512];
    char hashed_password[1024];
    char line[4096];
    char password[512];
    char salt[13];
    int found;

    if (!password_fp || !dictonary_fp)
    {
        perror("open():");
        exit(-1);
    }

    while (fgets(line, sizeof(line), password_fp) != NULL)
    {
        get_user_name_and_hashed_password(line, user, hashed_password);
        if (strcmp(hashed_password, "*") != 0 && strcmp(hashed_password, "!") != 0)
        {
            fprintf(stdout, "[*] Cracking password for %s...\n", user);
            found = 0;
            while (fscanf(dictonary_fp, "%s", password) != EOF && !found)
            {
                strncpy(salt, hashed_password, 12);
                salt[12] = '\0';
                //fprintf(stdout, "%s,%s,%s,%s", user, hashed_password, password, salt);
                //getchar();
                if (strcmp(crypt(password, salt), hashed_password) == 0)
                {
                    found = 1;
                    break;
                }
            }
            if (found)
            {
                fprintf(stdout, "\t[+] Password FOUND: %s\n", password);
            }
            else
            {
                fprintf(stdout, "\t[-] Password NOT FOUND\n");
            }
        }
        rewind(dictonary_fp);
    }

    fclose(password_fp);
    fclose(dictonary_fp);
}

void get_user_name_and_hashed_password(char *s, char *user, char *hashed_password)
{
    char *p = s;
    
    /* get user name */
    while (*p != ':')
    {
        *user++ = *p++;
    }
    *user = '\0';
    p++;

    /* get user's hashed password */
    while (*p != ':')
    {
        *hashed_password++ = *p++;
    }
    *hashed_password = '\0';

    return;
}

static inline void usage(char *prog)
{
    fprintf(stdout, "Usage: %s -f <password> -d <dictonary>\n", prog);
    return;
}